\name{alignStats}
\alias{alignStats}
\title{
Alignment statistics 
}
\description{
Generate data frame containg important read alignment statistics such as the
total number of reads in the FASTQ files, the number of total alignments, as
well as the number of primary alignments in the corresponding BAM files. 
}
\usage{
alignStats(fqpaths, bampaths, fqgz = TRUE)
}
\arguments{
  \item{fqpaths}{
	Paths to FASTQ files which can be obtained from the inflile1/2 component(s) of the list object generated by \code{systemArgs}, e.g. \code{fqpaths=tophatargs$infile1}.
}
  \item{bampaths}{
	Paths to BAM files which can be obtained from the output of the \code{runX} functions, e.g. \code{bampaths}.
}
  \item{fqgz}{
	Defines whether FASTQ files are compressed (\code{TRUE} or \code{FALSE}).
}
}
\details{
...
}
\value{
	\code{data.frame} with alignment statitics.
}
\references{
...
}
\author{
Thomas Girke
}
\note{
...
}
\seealso{
...
}
\examples{
\dontrun{
## Generate input targets file. Note: for \code{qsubRun()} the file targets_run.txt 
## needs to contain absolute paths to FASTQ files in "FileName' column.
library(systemPipeR)
targets <- read.delim("targets.txt", comment.char = "#")
write.table(targets, "targets_run.txt", row.names=FALSE, quote=FALSE, sep="\t")

## Build Bowtie 2 index
system("bowtie2-build ./data/mygenome.fa ./data/bowtie2index/mygenome")

## Run as single process without submitting to cluster, e.g. via qsub -I
mymodules <- c("bowtie2/2.1.0", "tophat/2.0.8b")
myargs <- c(software="tophat", p="-p 4", g="-g 1", segment_length="--segment-length 25", i="-i 30", I="-I 3000")
myref <- "./data/My_genome.fasta"
tophatargs <- systemArgs(app="tophat2", mymodules=mymodules, mydir=getwd(), myargs=myargs, myref=myref, mygff="My_specie.gff", mytargets="targets_run.txt")
bampaths <- runTophat(tophatargs=tophatargs, runid="01")

## Submit to compute nodes
qsubargs <- getQsubargs(queue="batch", Nnodes="nodes=4", cores=as.numeric(gsub("^.* ", "", tophatargs$args["p"])), memory="mem=10gb", time="walltime=20:00:00")
(joblist <- qsubRun(appfct="runTophat(appargs, runid)", appargs=tophatargs, qsubargs=qsubargs, Nqsubs=6, submitdir="/bigdata/tgirke/Projects/project_name/RNAseq/data", myfct="~/Projects/project_name/RNA-Seq/systemPipe.R"))

## Alignment stats
read_statsDF <- alignStats(fqpaths=tophatargs$infile1, bampaths=bampaths, fqgz=TRUE) 
read_statsDF <- cbind(read_statsDF[targets$FileName,], targets)
write.table(read_statsDF, "results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="\t")

}
}
\keyword{ utilities }
